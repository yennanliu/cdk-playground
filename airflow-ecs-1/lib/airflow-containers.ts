import * as ecs from 'aws-cdk-lib/aws-ecs';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Duration } from 'aws-cdk-lib';

export interface AirflowContainerProps {
  taskDefinition: ecs.TaskDefinition;
  dbEndpointAddress: string;
  dbEndpointPort: string; 
  redisEndpointAddress: string;
  redisEndpointPort: string;
  webserverLogGroup: logs.LogGroup;
  schedulerLogGroup: logs.LogGroup;
  workerLogGroup: logs.LogGroup;
  triggerLogGroup: logs.LogGroup;
}

export class AirflowContainers {
  public readonly webserver: ecs.ContainerDefinition;
  public readonly scheduler: ecs.ContainerDefinition;
  public readonly worker: ecs.ContainerDefinition;
  public readonly triggerer: ecs.ContainerDefinition;

  constructor(props: AirflowContainerProps) {
    const commonEnv = {
      AIRFLOW__CORE__EXECUTOR: 'CeleryExecutor',
      AIRFLOW__DATABASE__SQL_ALCHEMY_CONN: `postgresql://postgres:postgres@${props.dbEndpointAddress}:${props.dbEndpointPort}/airflow`,
      AIRFLOW__CELERY__BROKER_URL: `redis://${props.redisEndpointAddress}:${props.redisEndpointPort}/0`,
      AIRFLOW__CELERY__RESULT_BACKEND: `db+postgresql://postgres:postgres@${props.dbEndpointAddress}:${props.dbEndpointPort}/airflow`,
      AIRFLOW_HOME: '/opt/airflow',
      AIRFLOW__CORE__DAGS_FOLDER: '/opt/airflow/dags',
      AIRFLOW__CORE__LOAD_EXAMPLES: 'false',
      AIRFLOW__WEBSERVER__RBAC: 'true',
    };

    // Add webserver container
    this.webserver = props.taskDefinition.addContainer('webserver', {
      image: ecs.ContainerImage.fromRegistry('apache/airflow:3.0.1'),
      command: ['webserver'],
      environment: {
        ...commonEnv,
        AIRFLOW__WEBSERVER__EXPOSE_CONFIG: 'true',
      },
      logging: ecs.LogDrivers.awsLogs({
        streamPrefix: 'webserver',
        logGroup: props.webserverLogGroup,
      }),
      portMappings: [{
        containerPort: 8080,
        hostPort: 8080,
      }],
      healthCheck: {
        command: ['CMD-SHELL', 'curl --fail http://localhost:8080/health || exit 1'],
        interval: Duration.seconds(30),
        timeout: Duration.seconds(5),
        retries: 3,
        startPeriod: Duration.seconds(60),
      },
      essential: true,
    });

    // Add scheduler container
    this.scheduler = props.taskDefinition.addContainer('scheduler', {
      image: ecs.ContainerImage.fromRegistry('apache/airflow:3.0.1'),
      command: ['scheduler'],
      environment: commonEnv,
      logging: ecs.LogDrivers.awsLogs({
        streamPrefix: 'scheduler',
        logGroup: props.schedulerLogGroup,
      }),
      essential: true,
    });

    // Add worker container
    this.worker = props.taskDefinition.addContainer('worker', {
      image: ecs.ContainerImage.fromRegistry('apache/airflow:3.0.1'),
      command: ['celery', 'worker'],
      environment: commonEnv,
      logging: ecs.LogDrivers.awsLogs({
        streamPrefix: 'worker',
        logGroup: props.workerLogGroup,
      }),
      essential: true,
    });

    // Add triggerer container
    this.triggerer = props.taskDefinition.addContainer('triggerer', {
      image: ecs.ContainerImage.fromRegistry('apache/airflow:3.0.1'),
      command: ['triggerer'],
      environment: commonEnv,
      logging: ecs.LogDrivers.awsLogs({
        streamPrefix: 'triggerer',
        logGroup: props.triggerLogGroup,
      }),
      essential: true,
    });
  }
} 