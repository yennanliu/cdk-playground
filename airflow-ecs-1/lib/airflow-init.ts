import * as ecs from 'aws-cdk-lib/aws-ecs';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as cloudwatch from 'aws-cdk-lib/aws-cloudwatch';
import { Duration } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface AirflowInitProps {
  scope: Construct;
  cluster: ecs.ICluster;
  taskDefinition: ecs.TaskDefinition;
  dbEndpointAddress: string;
  dbEndpointPort: string;
  securityGroups: ec2.SecurityGroup[];
  vpcSubnets: ec2.SubnetSelection;
  logGroup: logs.LogGroup;
}

export class AirflowInit {
  public readonly initTask: ecs.FargateService;
  
  constructor(props: AirflowInitProps) {
    // Create the init container
    const initContainer = props.taskDefinition.addContainer('init', {
      image: ecs.ContainerImage.fromRegistry('apache/airflow:3.0.1'),
      command: [
        'bash', 
        '-c', 
        'airflow db init && ' +
        'airflow users create --username airflow --password airflow --firstname Admin ' +
        '--lastname User --role Admin --email admin@example.com'
      ],
      environment: {
        AIRFLOW__CORE__EXECUTOR: 'LocalExecutor',
        AIRFLOW__DATABASE__SQL_ALCHEMY_CONN: `postgresql://postgres:postgres@${props.dbEndpointAddress}:${props.dbEndpointPort}/airflow`,
        AIRFLOW_HOME: '/opt/airflow',
        AIRFLOW__CORE__LOAD_EXAMPLES: 'false',
      },
      logging: ecs.LogDrivers.awsLogs({
        streamPrefix: 'init',
        logGroup: props.logGroup,
      }),
      essential: true,
    });

    // Create a one-time task to initialize Airflow
    this.initTask = new ecs.FargateService(props.scope, 'AirflowInitTask', {
      cluster: props.cluster,
      taskDefinition: props.taskDefinition,
      desiredCount: 0, // Run once and then stop
      securityGroups: props.securityGroups,
      vpcSubnets: props.vpcSubnets,
      assignPublicIp: false,
    });

    // Add auto-scaling policy to scale to 1 to run the init task once
    const scaling = this.initTask.autoScaleTaskCount({
      minCapacity: 0,
      maxCapacity: 1,
    });

    // Run the task once during deployment
    scaling.scaleToTrackCustomMetric('RunInitTask', {
      targetValue: 1,
      scaleInCooldown: Duration.seconds(60),
      scaleOutCooldown: Duration.seconds(60),
      metric: new cloudwatch.Metric({
        namespace: 'AWS/ECS',
        metricName: 'CPUUtilization',
        dimensionsMap: {
          ClusterName: props.cluster.clusterName,
          ServiceName: this.initTask.serviceName,
        },
        statistic: 'Average',
        period: Duration.seconds(60),
      }),
    });
  }
} 