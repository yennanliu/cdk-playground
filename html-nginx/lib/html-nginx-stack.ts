import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as s3deploy from 'aws-cdk-lib/aws-s3-deployment';
import * as assets from 'aws-cdk-lib/aws-s3-assets';
import * as autoscaling from 'aws-cdk-lib/aws-autoscaling';
import * as elbv2 from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import { Construct } from 'constructs';
import * as path from 'path';

export class HtmlNginxStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create a VPC
    const vpc = new ec2.Vpc(this, 'WebsiteVPC', {
      maxAzs: 2,
      natGateways: 0, // To save costs, we won't use NAT gateways
    });

    // Create a security group for the web server
    const webSecurityGroup = new ec2.SecurityGroup(this, 'WebSecurityGroup', {
      vpc,
      description: 'Allow HTTP access to web servers',
      allowAllOutbound: true,
    });
    webSecurityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(80), 'Allow HTTP access from anywhere');

    // Create a role for the instances
    const webServerRole = new iam.Role(this, 'WebServerRole', {
      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
    });
    webServerRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore'));
    
    // Upload website content to S3
    const websiteAsset = new assets.Asset(this, 'WebsiteContent', {
      path: path.join(__dirname, '../website'),
    });
    
    // Create nginx.conf as an asset
    const nginxConfigAsset = new assets.Asset(this, 'NginxConfig', {
      path: path.join(__dirname, '../website/nginx.conf'),
    });

    // Create user data for EC2 instances
    const userData = ec2.UserData.forLinux();
    userData.addCommands(
      'yum update -y',
      'amazon-linux-extras install nginx1 -y',
      'yum install -y aws-cli',
      // Create directories
      'mkdir -p /var/www/html',
      // Download website files from S3
      `aws s3 cp ${websiteAsset.s3ObjectUrl} /tmp/website.zip`,
      'cd /var/www/html && unzip /tmp/website.zip',
      // Download Nginx config
      `aws s3 cp ${nginxConfigAsset.s3ObjectUrl} /etc/nginx/nginx.conf`,
      // Set proper permissions
      'chmod -R 755 /var/www/html',
      'chown -R nginx:nginx /var/www/html',
      // Start Nginx
      'systemctl start nginx',
      'systemctl enable nginx'
    );

    // Create auto scaling group
    const autoScalingGroup = new autoscaling.AutoScalingGroup(this, 'WebServerASG', {
      vpc,
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.T2, ec2.InstanceSize.MICRO),
      machineImage: ec2.MachineImage.latestAmazonLinux2({
        cpuType: ec2.AmazonLinuxCpuType.X86_64,
      }),
      securityGroup: webSecurityGroup,
      userData,
      role: webServerRole,
      minCapacity: 2,
      maxCapacity: 4,
      desiredCapacity: 2,
    });

    // Grant read access to S3 assets
    websiteAsset.grantRead(autoScalingGroup.role);
    nginxConfigAsset.grantRead(autoScalingGroup.role);

    // Create application load balancer
    const alb = new elbv2.ApplicationLoadBalancer(this, 'WebLoadBalancer', {
      vpc,
      internetFacing: true,
    });

    // Create listener and target group
    const listener = alb.addListener('HttpListener', {
      port: 80,
      open: true,
    });

    listener.addTargets('WebTargetGroup', {
      port: 80,
      targets: [autoScalingGroup],
      healthCheck: {
        path: '/',
        interval: cdk.Duration.seconds(60),
        timeout: cdk.Duration.seconds(5),
      },
    });

    // Output the load balancer URL
    new cdk.CfnOutput(this, 'LoadBalancerURL', {
      value: `http://${alb.loadBalancerDnsName}`,
      description: 'URL of the load balancer',
    });
  }
}
