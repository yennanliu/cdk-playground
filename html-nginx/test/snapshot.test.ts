import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as HtmlNginx from '../lib/html-nginx-stack';

test('Stack creates snapshot', () => {
  const app = new cdk.App();
  const stack = new HtmlNginx.HtmlNginxStack(app, 'SnapshotHtmlNginxStack');
  const template = Template.fromStack(stack);
  
  // Convert the template to a JSON string and match it to a stored snapshot
  expect(JSON.stringify(template.toJSON(), null, 2)).toMatchSnapshot();
}); 