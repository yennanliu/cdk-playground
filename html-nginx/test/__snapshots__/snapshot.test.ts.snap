// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack creates snapshot 1`] = `
"{
  "Resources": {
    "WebsiteVPCD8B49DC8": {
      "Type": "AWS::EC2::VPC",
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "SnapshotHtmlNginxStack/WebsiteVPC"
          }
        ]
      }
    },
    "WebsiteVPCPublicSubnet1Subnet2406F279": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": ""
            }
          ]
        },
        "CidrBlock": "10.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public"
          },
          {
            "Key": "Name",
            "Value": "SnapshotHtmlNginxStack/WebsiteVPC/PublicSubnet1"
          }
        ],
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    },
    "WebsiteVPCPublicSubnet1RouteTable6301BDB8": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "SnapshotHtmlNginxStack/WebsiteVPC/PublicSubnet1"
          }
        ],
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    },
    "WebsiteVPCPublicSubnet1RouteTableAssociation6ABF5A2B": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "WebsiteVPCPublicSubnet1RouteTable6301BDB8"
        },
        "SubnetId": {
          "Ref": "WebsiteVPCPublicSubnet1Subnet2406F279"
        }
      }
    },
    "WebsiteVPCPublicSubnet1DefaultRouteC187236C": {
      "Type": "AWS::EC2::Route",
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "WebsiteVPCIGWD7296D39"
        },
        "RouteTableId": {
          "Ref": "WebsiteVPCPublicSubnet1RouteTable6301BDB8"
        }
      },
      "DependsOn": [
        "WebsiteVPCVPCGW9965E50A"
      ]
    },
    "WebsiteVPCPublicSubnet2Subnet8E2CB9C8": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": ""
            }
          ]
        },
        "CidrBlock": "10.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public"
          },
          {
            "Key": "Name",
            "Value": "SnapshotHtmlNginxStack/WebsiteVPC/PublicSubnet2"
          }
        ],
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    },
    "WebsiteVPCPublicSubnet2RouteTable3A89D35C": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "SnapshotHtmlNginxStack/WebsiteVPC/PublicSubnet2"
          }
        ],
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    },
    "WebsiteVPCPublicSubnet2RouteTableAssociation95466B5A": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "WebsiteVPCPublicSubnet2RouteTable3A89D35C"
        },
        "SubnetId": {
          "Ref": "WebsiteVPCPublicSubnet2Subnet8E2CB9C8"
        }
      }
    },
    "WebsiteVPCPublicSubnet2DefaultRoute5760AF69": {
      "Type": "AWS::EC2::Route",
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "WebsiteVPCIGWD7296D39"
        },
        "RouteTableId": {
          "Ref": "WebsiteVPCPublicSubnet2RouteTable3A89D35C"
        }
      },
      "DependsOn": [
        "WebsiteVPCVPCGW9965E50A"
      ]
    },
    "WebsiteVPCIsolatedSubnet1Subnet155D5221": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": ""
            }
          ]
        },
        "CidrBlock": "10.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated"
          },
          {
            "Key": "Name",
            "Value": "SnapshotHtmlNginxStack/WebsiteVPC/IsolatedSubnet1"
          }
        ],
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    },
    "WebsiteVPCIsolatedSubnet1RouteTable0D7FA7EB": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "SnapshotHtmlNginxStack/WebsiteVPC/IsolatedSubnet1"
          }
        ],
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    },
    "WebsiteVPCIsolatedSubnet1RouteTableAssociation62E83A89": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "WebsiteVPCIsolatedSubnet1RouteTable0D7FA7EB"
        },
        "SubnetId": {
          "Ref": "WebsiteVPCIsolatedSubnet1Subnet155D5221"
        }
      }
    },
    "WebsiteVPCIsolatedSubnet2SubnetD62B6F84": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": ""
            }
          ]
        },
        "CidrBlock": "10.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated"
          },
          {
            "Key": "Name",
            "Value": "SnapshotHtmlNginxStack/WebsiteVPC/IsolatedSubnet2"
          }
        ],
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    },
    "WebsiteVPCIsolatedSubnet2RouteTable991E86C1": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "SnapshotHtmlNginxStack/WebsiteVPC/IsolatedSubnet2"
          }
        ],
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    },
    "WebsiteVPCIsolatedSubnet2RouteTableAssociation35362FAF": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "WebsiteVPCIsolatedSubnet2RouteTable991E86C1"
        },
        "SubnetId": {
          "Ref": "WebsiteVPCIsolatedSubnet2SubnetD62B6F84"
        }
      }
    },
    "WebsiteVPCIGWD7296D39": {
      "Type": "AWS::EC2::InternetGateway",
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "SnapshotHtmlNginxStack/WebsiteVPC"
          }
        ]
      }
    },
    "WebsiteVPCVPCGW9965E50A": {
      "Type": "AWS::EC2::VPCGatewayAttachment",
      "Properties": {
        "InternetGatewayId": {
          "Ref": "WebsiteVPCIGWD7296D39"
        },
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    },
    "WebSecurityGroup73AF7387": {
      "Type": "AWS::EC2::SecurityGroup",
      "Properties": {
        "GroupDescription": "Allow HTTP access to web servers",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1"
          }
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow HTTP access from anywhere",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80
          }
        ],
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    },
    "WebSecurityGroupfromSnapshotHtmlNginxStackWebLoadBalancerSecurityGroupFA26520F807E48B333": {
      "Type": "AWS::EC2::SecurityGroupIngress",
      "Properties": {
        "Description": "Load balancer to target",
        "FromPort": 80,
        "GroupId": {
          "Fn::GetAtt": [
            "WebSecurityGroup73AF7387",
            "GroupId"
          ]
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "WebLoadBalancerSecurityGroup791DED1A",
            "GroupId"
          ]
        },
        "ToPort": 80
      }
    },
    "WebServerRole42ABA571": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore"
              ]
            ]
          }
        ]
      }
    },
    "WebServerRoleDefaultPolicy9AFE19CE": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}"
                      }
                    ]
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}"
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "WebServerRoleDefaultPolicy9AFE19CE",
        "Roles": [
          {
            "Ref": "WebServerRole42ABA571"
          }
        ]
      }
    },
    "WebServerASGInstanceProfile1031076F": {
      "Type": "AWS::IAM::InstanceProfile",
      "Properties": {
        "Roles": [
          {
            "Ref": "WebServerRole42ABA571"
          }
        ]
      }
    },
    "WebServerASGLaunchConfig965C5C3F": {
      "Type": "AWS::AutoScaling::LaunchConfiguration",
      "Properties": {
        "IamInstanceProfile": {
          "Ref": "WebServerASGInstanceProfile1031076F"
        },
        "ImageId": {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amikernel510hvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter"
        },
        "InstanceType": "t2.micro",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "WebSecurityGroup73AF7387",
              "GroupId"
            ]
          }
        ],
        "UserData": {
          "Fn::Base64": {
            "Fn::Join": [
              "",
              [
                "#!/bin/bash\\nyum update -y\\namazon-linux-extras install nginx1 -y\\nyum install -y aws-cli\\nmkdir -p /var/www/html\\naws s3 cp ",
                {
                  "Fn::Sub": "s3://cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}/497b21ebdb771071ba64a314659780b2a44d930a207513736cdc44dcbe28f4a7.zip"
                },
                " /tmp/website.zip\\ncd /var/www/html && unzip /tmp/website.zip\\naws s3 cp ",
                {
                  "Fn::Sub": "s3://cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}/4daf42ffaa1611f7878dcab4220e174bf52b519f21f6ce56eab91376dc849e6e.conf"
                },
                " /etc/nginx/nginx.conf\\nchmod -R 755 /var/www/html\\nchown -R nginx:nginx /var/www/html\\nsystemctl start nginx\\nsystemctl enable nginx"
              ]
            ]
          }
        }
      },
      "DependsOn": [
        "WebServerRoleDefaultPolicy9AFE19CE",
        "WebServerRole42ABA571"
      ]
    },
    "WebServerASG5A722E5A": {
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "Properties": {
        "DesiredCapacity": "2",
        "LaunchConfigurationName": {
          "Ref": "WebServerASGLaunchConfig965C5C3F"
        },
        "MaxSize": "4",
        "MinSize": "2",
        "Tags": [
          {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "SnapshotHtmlNginxStack/WebServerASG"
          }
        ],
        "TargetGroupARNs": [
          {
            "Ref": "WebLoadBalancerHttpListenerWebTargetGroupGroup159A7108"
          }
        ],
        "VPCZoneIdentifier": [
          {
            "Ref": "WebsiteVPCIsolatedSubnet1Subnet155D5221"
          },
          {
            "Ref": "WebsiteVPCIsolatedSubnet2SubnetD62B6F84"
          }
        ]
      },
      "UpdatePolicy": {
        "AutoScalingScheduledAction": {
          "IgnoreUnmodifiedGroupSizeProperties": true
        }
      }
    },
    "WebLoadBalancer3B0ADE5E": {
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
      "Properties": {
        "LoadBalancerAttributes": [
          {
            "Key": "deletion_protection.enabled",
            "Value": "false"
          }
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "WebLoadBalancerSecurityGroup791DED1A",
              "GroupId"
            ]
          }
        ],
        "Subnets": [
          {
            "Ref": "WebsiteVPCPublicSubnet1Subnet2406F279"
          },
          {
            "Ref": "WebsiteVPCPublicSubnet2Subnet8E2CB9C8"
          }
        ],
        "Type": "application"
      },
      "DependsOn": [
        "WebsiteVPCPublicSubnet1DefaultRouteC187236C",
        "WebsiteVPCPublicSubnet1RouteTableAssociation6ABF5A2B",
        "WebsiteVPCPublicSubnet2DefaultRoute5760AF69",
        "WebsiteVPCPublicSubnet2RouteTableAssociation95466B5A"
      ]
    },
    "WebLoadBalancerSecurityGroup791DED1A": {
      "Type": "AWS::EC2::SecurityGroup",
      "Properties": {
        "GroupDescription": "Automatically created Security Group for ELB SnapshotHtmlNginxStackWebLoadBalancer3AD66B64",
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80
          }
        ],
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    },
    "WebLoadBalancerSecurityGrouptoSnapshotHtmlNginxStackWebSecurityGroupF10FD2B780839ADA70": {
      "Type": "AWS::EC2::SecurityGroupEgress",
      "Properties": {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": {
          "Fn::GetAtt": [
            "WebSecurityGroup73AF7387",
            "GroupId"
          ]
        },
        "FromPort": 80,
        "GroupId": {
          "Fn::GetAtt": [
            "WebLoadBalancerSecurityGroup791DED1A",
            "GroupId"
          ]
        },
        "IpProtocol": "tcp",
        "ToPort": 80
      }
    },
    "WebLoadBalancerHttpListener1488FDD1": {
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
      "Properties": {
        "DefaultActions": [
          {
            "TargetGroupArn": {
              "Ref": "WebLoadBalancerHttpListenerWebTargetGroupGroup159A7108"
            },
            "Type": "forward"
          }
        ],
        "LoadBalancerArn": {
          "Ref": "WebLoadBalancer3B0ADE5E"
        },
        "Port": 80,
        "Protocol": "HTTP"
      }
    },
    "WebLoadBalancerHttpListenerWebTargetGroupGroup159A7108": {
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
      "Properties": {
        "HealthCheckIntervalSeconds": 60,
        "HealthCheckPath": "/",
        "HealthCheckTimeoutSeconds": 5,
        "Port": 80,
        "Protocol": "HTTP",
        "TargetGroupAttributes": [
          {
            "Key": "stickiness.enabled",
            "Value": "false"
          }
        ],
        "TargetType": "instance",
        "VpcId": {
          "Ref": "WebsiteVPCD8B49DC8"
        }
      }
    }
  },
  "Parameters": {
    "SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amikernel510hvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
      "Default": "/aws/service/ami-amazon-linux-latest/amzn2-ami-kernel-5.10-hvm-x86_64-gp2"
    },
    "BootstrapVersion": {
      "Type": "AWS::SSM::Parameter::Value<String>",
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]"
    }
  },
  "Outputs": {
    "LoadBalancerURL": {
      "Description": "URL of the load balancer",
      "Value": {
        "Fn::Join": [
          "",
          [
            "http://",
            {
              "Fn::GetAtt": [
                "WebLoadBalancer3B0ADE5E",
                "DNSName"
              ]
            }
          ]
        ]
      }
    }
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5"
                  ],
                  {
                    "Ref": "BootstrapVersion"
                  }
                ]
              }
            ]
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI."
        }
      ]
    }
  }
}"
`;
